Multiprocessing incompatibility with find_essential_genes

Hi!

When running the cameo in methods using multiprocessing, finding essential genes is broken.

cameo (0.11.3)
cobra (0.10.1)

One fix can be to allow one to set the number of jobs to one in the find_essential_genes method, so it wouldn't run using multiprocessing. Because when the `num_jobs` argument is None, cobrapy takes the `cpu_count` and computes gene deletions using a Pool. However, it is not possible to spawn a pool inside a child process.

The fix would be to check allow setting the number of jobs and run single process mode in cameo. Computing the essential genes is not the bottleneck in any of the methods.

Cheers,

João

Code cameo-notebooks/08-high-level-API.ipynb

Traceback:
------------
```
Optimizing 8 pathways
---------------------------------------------------------------------------
RemoteTraceback                           Traceback (most recent call last)
RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cameo/api/designer.py", line 94, in __call__
    max_evaluations=max_evaluations, max_knockouts=15, max_time=max_time)
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cameo/strain_design/heuristic/evolutionary_based.py", line 144, in run
    use_nullspace_simplification=use_nullspace_simplification)
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cameo/strain_design/heuristic/evolutionary/optimization.py", line 766, in __init__
    self.essential_genes = {g.id for g in find_essential_genes(self.model)}
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cobra/flux_analysis/variability.py", line 240, in find_essential_genes
    deletions = single_gene_deletion(model, method='fba')
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cobra/flux_analysis/deletion.py", line 263, in single_gene_deletion
    method=method, num_jobs=num_jobs)
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cobra/flux_analysis/deletion.py", line 147, in _multi_deletion
    num_cpu, initializer=_init_worker, initargs=(model,)
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/multiprocessing/context.py", line 118, in Pool
    context=self.get_context())
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/multiprocessing/pool.py", line 168, in __init__
    self._repopulate_pool()
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/multiprocessing/pool.py", line 233, in _repopulate_pool
    w.start()
  File "/usr/local/miniconda3/envs/cameo/lib/python3.4/multiprocessing/process.py", line 103, in start
    'daemonic processes are not allowed to have children'
AssertionError: daemonic processes are not allowed to have children
"""

The above exception was the direct cause of the following exception:

AssertionError                            Traceback (most recent call last)
<ipython-input-8-327cca7e82b0> in <module>()
----> 1 report = api.design(product='vanillin', view=mp_view)

/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cameo/api/designer.py in __call__(self, product, hosts, database, aerobic, view)
    185             raise KeyError("Product %s is not in the %s database" % (product, database.id))
    186         pathways = self.predict_pathways(product, hosts=hosts, database=database, aerobic=aerobic)
--> 187         optimization_reports = self.optimize_strains(pathways, view, aerobic=aerobic)
    188         return optimization_reports
    189 

/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cameo/api/designer.py in optimize_strains(self, pathways, view, aerobic)
    219             logger.info('Running heuristic optimizations on predicted pathways.')
    220             opt_gene_runner = _OptGeneRunner(options=self.options)
--> 221             designs = view.map(opt_gene_runner, strategies)
    222             logger.info('Processing heuristic optimization results.')
    223             results = self.build_results_data(designs, strategies, results, 'PathwayPredictor+OptGene')

/usr/local/miniconda3/envs/cameo/lib/python3.4/site-packages/cameo/parallel.py in map(self, *args, **kwargs)
     53 
     54     def map(self, *args, **kwargs):
---> 55         return self.pool.map(*args, **kwargs)
     56 
     57     def apply(self, func, *args, **kwargs):

/usr/local/miniconda3/envs/cameo/lib/python3.4/multiprocessing/pool.py in map(self, func, iterable, chunksize)
    258         in a list that is returned.
    259         '''
--> 260         return self._map_async(func, iterable, mapstar, chunksize).get()
    261 
    262     def starmap(self, func, iterable, chunksize=None):

/usr/local/miniconda3/envs/cameo/lib/python3.4/multiprocessing/pool.py in get(self, timeout)
    606             return self._value
    607         else:
--> 608             raise self._value
    609 
    610     def _set(self, i, obj):

AssertionError: daemonic processes are not allowed to have children
```

--- a/cameo/strain_design/deterministic/linear_programming.py
+++ b/cameo/strain_design/deterministic/linear_programming.py
@@ -131,7 +131,7 @@ class OptKnock(StrainDesignMethod):
     def _build_problem(self, essential_reactions, use_nullspace_simplification):
         logger.debug("Starting to formulate OptKnock problem")
 
-        self.essential_reactions = find_essential_reactions(self._model).union(self._model.exchanges)
+        self.essential_reactions = find_essential_reactions(self._model, processes=1).union(self._model.exchanges)
         if essential_reactions:
             self.essential_reactions.update(set(get_reaction_for(self._model, r) for r in essential_reactions))
 

--- a/cameo/strain_design/heuristic/evolutionary/multiprocess/optimization.py
+++ b/cameo/strain_design/heuristic/evolutionary/multiprocess/optimization.py
@@ -214,7 +214,7 @@ class MultiprocessReactionKnockoutOptimization(MultiprocessKnockoutOptimization)
             self.reactions = reactions
 
         if essential_reactions is None:
-            self.essential_reactions = set([r.id for r in find_essential_reactions(self.model)])
+            self.essential_reactions = set([r.id for r in find_essential_reactions(self.model, processes=1)])
         else:
             self.essential_reactions = essential_reactions
 
@@ -257,7 +257,7 @@ class MultiprocessGeneKnockoutOptimization(MultiprocessKnockoutOptimization):
             self.genes = genes
 
         if essential_genes is None:
-            self.essential_genes = set([g.id for g in find_essential_genes(self.model)])
+            self.essential_genes = set([g.id for g in find_essential_genes(self.model, processes=1)])
         else:
             self.essential_genes = essential_genes
 

--- a/cameo/strain_design/heuristic/evolutionary/optimization.py
+++ b/cameo/strain_design/heuristic/evolutionary/optimization.py
@@ -670,9 +670,10 @@ class ReactionKnockoutOptimization(KnockoutOptimization):
             self.reactions = reactions
         logger.debug("Computing essential reactions...")
         if essential_reactions is None:
-            self.essential_reactions = set(r.id for r in find_essential_reactions(self.model))
+            self.essential_reactions = set(r.id for r in find_essential_reactions(self.model, processes=1))
         else:
-            self.essential_reactions = set([r.id for r in find_essential_reactions(self.model)] + essential_reactions)
+            self.essential_reactions = set([r.id for r in find_essential_reactions(self.model, processes=1)]
+                                           + essential_reactions)
 
         if use_nullspace_simplification:
             ns = nullspace(create_stoichiometric_array(self.model))
@@ -763,9 +764,9 @@ class GeneKnockoutOptimization(KnockoutOptimization):
         else:
             self.genes = genes
         if essential_genes is None:
-            self.essential_genes = {g.id for g in find_essential_genes(self.model)}
+            self.essential_genes = {g.id for g in find_essential_genes(self.model, processes=1)}
         else:
-            self.essential_genes = set([g.id for g in find_essential_genes(self.model)] + essential_genes)
+            self.essential_genes = set([g.id for g in find_essential_genes(self.model, processes=1)] + essential_genes)
 
         # TODO: use genes from groups
         if use_nullspace_simplification:

--- a/cameo/strain_design/heuristic/evolutionary/optimization.py
+++ b/cameo/strain_design/heuristic/evolutionary/optimization.py
@@ -672,8 +672,8 @@ class ReactionKnockoutOptimization(KnockoutOptimization):
         if essential_reactions is None:
             self.essential_reactions = set(r.id for r in find_essential_reactions(self.model, processes=1))
         else:
-            self.essential_reactions = set([r.id for r in find_essential_reactions(self.model, processes=1)]
-                                           + essential_reactions)
+            self.essential_reactions = set([r.id for r in find_essential_reactions(self.model, processes=1)])
+            self.essential_reactions.update(essential_reactions)
 
         if use_nullspace_simplification:
             ns = nullspace(create_stoichiometric_array(self.model))

